pulumi/src/coinstack.ts

```diff
@@ -0,0 +1,145 @@
+): Promise<Outputs> => {
+  const { kubeconfig, config, namespace } = await getConfig()
+
+  const assetName = config.network !== 'mainnet' ? `${config.assetName}-${config.network}` : config.assetName
+  const provider = new k8s.Provider('kube-provider', { kubeconfig })
+  const secretData = getSecretData(sampleEnv)
+
+  new k8s.core.v1.Secret(assetName, { metadata: { name: assetName, namespace }, stringData: secretData }, { provider })
+
+  const baseImageName = 'shapeshiftdao/unchained-base:latest'
+  const snapshots = await new Snapper({ assetName, kubeconfig, namespace }).getSnapshots()
+
+  await deployApi({
+    appName,
+    assetName,
+    coinstack,
+    coinstackType,
+    sampleEnv,
+    baseImageName,
+    config,
+    namespace,
+    provider,
+  })
+
+  const coinServices = aggregateCoinServiceInput(config.statefulService?.services || [], serviceInput)
+    .map((cs) => enrichWithPulumiConfig(cs, namespace, coinstack, config))
+    .map((cs) => createService(cs, assetName, snapshots))
+  await deployStatefulService(appName, assetName, provider, namespace, config, coinServices, volumes)
+  return {}
+}
+
+const enrichWithPulumiConfig = (
+  csi: JointCoinServiceInput,
+  namespace: string,
+  coinstack: string,
+  config: Config
+): JointCoinServiceInput => {
+  // TODO how to solve this elegantly so that we know what to apply to which service?
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'thorchain') {
+    csi.env = {
+      ...csi.env,
+      CHAIN_ID: `${coinstack}-${config.network}-v1`,
+      NET: config.network,
+    }
+  }
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'ethereum') {
+    csi.env = {
+      ...csi.env,
+      NETWORK: config.network,
+    }
+  }
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'bnbsmartchain') {
+    csi.env = {
+      ...csi.env,
+      SNAPSHOT: 'https://pub-c0627345c16f47ab858c9469133073a8.r2.dev/geth-20230409.tar.lz4',
+    }
+  }
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'dogecoin') {
+    csi.env = {
+      ...csi.env,
+      NETWORK: config.network,
+    }
+  }
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'bitcoin') {
+    csi.env = {
+      ...csi.env,
+      NETWORK: config.network,
+    }
+  }
+
+  if (csi.coinServiceName == 'daemon' && coinstack == 'bitcoincash') {
+    csi.env = {
+      ...csi.env,
+      NETWORK: config.network,
+    }
+  }
+
+  if (csi.coinServiceName == 'heimdall' && coinstack == 'polygon') {
+    csi.env = {
+      ...csi.env,
+      ETH_RPC_URL: `http://ethereum-svc.${namespace}.svc.cluster.local:8332`,
+    }
+  }
+
+  if (csi.coinServiceName == 'dtl' && coinstack == 'optimism') {
+    csi.env = {
+      ...csi.env,
+      L1_RPC_ENDPOINT: `http://ethereum-svc.${namespace}.svc.cluster.local:8332`,
+    }
+  }
+
+  return csi
+}
+
+// Join the config set in the Pulumi config with the typescript input from the coinstack
+const aggregateCoinServiceInput = (
```

Comment:
wont need this by lifting config either

Reply:
I did the lift but I think this needs to stay. Can't have a unified type for both of these, both have unique required args